tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -38,131 +40,161 @@ import           Source.Span
 --
---   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'unmarshalNode' providing that they have a suitable 'Generic' instance.
-class Unmarshal a where
-
-  unmarshalNodes :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
-  default unmarshalNodes :: (MonadFail m, Carrier sig m, GUnmarshal (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
-  unmarshalNodes [x] = do
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'unmarshalNode' providing that they have a suitable 'Generic1' instance.
+class Unmarshal t where
+  unmarshalNode
+    :: ( Carrier sig m
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       , UnmarshalAnn a
+       )
+    => Node
+    -> m (t a)
+  default unmarshalNode
+    :: ( Carrier sig m
+       , Generic1 t
+       , GUnmarshal (Rep1 t)
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       , UnmarshalAnn a
+       )
+    => Node
+    -> m (t a)
+  unmarshalNode x = do
     goto (nodeTSNode x)
-    to <$> gunmarshalNode x
-  unmarshalNodes [] = fail "expected a node but didn't get one"
-  unmarshalNodes _ = fail "expected a node but got multiple"
+    to1 <$> gunmarshalNode x
 
-instance Unmarshal () where
-  unmarshalNodes _ = pure ()
+instance (Unmarshal f, Unmarshal g, SymbolMatching f, SymbolMatching g) => Unmarshal (f :+: g) where
+  unmarshalNode node = do
+    let lhsSymbolMatch = symbolMatch (Proxy @f) node
+        rhsSymbolMatch = symbolMatch (Proxy @g) node
+    if lhsSymbolMatch then
+      L1 <$> unmarshalNode @f node
+    else if rhsSymbolMatch then
+      R1 <$> unmarshalNode @g node
+    else
+      fail $ showFailure (Proxy @(f :+: g)) node
 
-instance Unmarshal Text.Text where
-  unmarshalNodes _ = do
-    node <- peekNode
+instance Unmarshal t => Unmarshal (Rec1 t) where
+  unmarshalNode = fmap Rec1 . unmarshalNode
+
+
+-- | Unmarshal an annotation field.
+--
+--   Leaf nodes have 'Text.Text' fields, and leaves, anonymous leaves, and products all have parametric annotation fields. All of these fields are unmarshalled using the metadata of the node, e.g. its start/end bytes, without reference to any child nodes it may contain.
+class UnmarshalAnn a where
+  unmarshalAnn
+    :: ( Carrier sig m
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       )
+    => Node
+    -> m a
+
+instance UnmarshalAnn () where
+  unmarshalAnn _ = pure ()
+
+instance UnmarshalAnn Text.Text where
+  unmarshalAnn node = do
```

Comment:
Because `unmarshalAnn` takes the node instead of having to `peekNode`, it’s much simpler and doesn’t have to deal with the possibility of failure. Happily, this also entails less crossing of the FFI!